lib/dispatcher.js

```diff
@@ -0,0 +1,187 @@
+  }
+
+  serverStarting() {
+    this.#makeRequest(
+      { requestType: 'server-starting' },
+      `${this.#url}/internal/brokerservers/${this.#id}`,
+      'post',
+      { health_check_link: `http://${this.#hostname}/healthcheck` },
+    );
+  }
+
+  serverStopping(cb) {
+    this.#makeRequest(
+      { requestType: 'server-stopping' },
+      `${this.#url}/internal/brokerservers/${this.#id}`,
+      'delete',
+      cb,
+    );
+  }
+
+  clientConnected(token, clientId) {
+    const maskedToken = maskToken(token);
+    this.#makeRequest(
+      { maskedToken, clientId, requestType: 'client-connected' },
+      `${this.#url}/internal/brokerservers/${this.#id}/connections/${hash(
+        token,
+      )}${clientId ? `?brokerClientId=${clientId}` : ''}`,
+      'post',
+      { health_check_link: `http://${this.#hostname}/healthcheck` },
+    );
+  }
+
+  clientDisconnected(token, clientId) {
+    const maskedToken = maskToken(token);
+    this.#makeRequest(
+      { maskedToken, clientId, requestType: 'client-disconnected' },
+      `${this.#url}/internal/brokerservers/${this.#id}/connections/${hash(
+        token,
+      )}${clientId ? `?brokerClientId=${clientId}` : ''}`,
+      'delete',
+    );
+  }
+
+  #makeRequest(logContext, url, method, requestBody, cb) {
+    // version *must* be provided
+    url = `${url}?version=${this.#version}`;
+    let body = '';
+    let statusCode = -1;
+    let headers = {};
+    request({
+      url,
+      method,
+      headers: {
+        'Content-Type': 'application/json',
+        Connection: 'Keep-Alive',
+        'Keep-Alive': 'timeout=60, max=10',
+      },
+      body: JSON.stringify(requestBody),
+    })
+      .on('error', (e) => {
+        logger.error(
+          {
+            ...logContext,
+            error: e,
+            dispatcherUrl: this.#url,
+            dispatcherVersion: this.#version,
+            id: this.#id,
+          },
+          'received error communicating with Dispatcher',
+        );
+        if (cb) cb();
+      })
+      .on('response', (r) => {
+        statusCode = r.statusCode;
+        headers = r.headers;
+      })
+      .on('data', (d) => {
+        body += d.toString();
+      })
+      .on('end', () => {
+        if (statusCode >= 300) {
+          logger.error(
+            {
+              ...logContext,
+              statusCode,
+              headers,
+              body,
+              dispatcherUrl: this.#url,
+              dispatcherVersion: this.#version,
+              id: this.#id,
+            },
+            'received unexpected status code communicating with Dispatcher',
+          );
+        } else {
+          logger.trace(
+            { ...logContext, id: this.#id },
+            'successfully sent request to Dispatcher',
+          );
+        }
+        if (cb) cb();
+      });
+  }
+}
+
+function hash(token) {
+  const shasum = crypto.createHash('sha256');
+  shasum.update(token);
+  return shasum.digest('hex');
+}
```

Comment:
@ipsi, it makes sense to export this function, so I can re-use it as well on the client side.